File: firebase/functions/lib/admin/payments/create_donation_checkout_session.dart
import 'dart:math';

import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import '../../on_call_function.dart';
import '../../utils/infra/firestore_utils.dart';
import 'stripe_util.dart';
import '../../utils/subscription_plan_util.dart';
import '../../utils/utils.dart';
import 'package:data_models/cloud_functions/requests.dart';
import 'package:data_models/community/community.dart';
import 'package:data_models/admin/partner_agreement.dart';

class CreateDonationCheckoutSession
    extends OnCallMethod<CreateDonationCheckoutSessionRequest> {
  CreateDonationCheckoutSession()
      : super(
          'createDonationCheckoutSession',
          (jsonMap) => CreateDonationCheckoutSessionRequest.fromJson(jsonMap),
        );

  @override
  Future<Map<String, dynamic>> action(
    CreateDonationCheckoutSessionRequest request,
    CallableContext context,
  ) async {
    final amountInCents = request.amountInCents;
    if (amountInCents <= 0) {
      throw HttpsError(
        HttpsError.invalidArgument,
        'invalid argument',
        'AmountInCents must be greater than zero.',
      );
    }

    final communityId = request.communityId;
    final orgName = functions.config.get('app.legal_entity_name') as String;

    final String customerId =
        await stripeUtil.getOrCreateCustomerStripeId(uid: context.authUid!);
    final domain = functions.config.get('app.domain') as String;

    final communitySnapshot =
        await firestore.document('/community/$communityId').get();
    orElseNotFound(communitySnapshot.exists);
    final community = Community.fromJson(
      firestoreUtils.fromFirestoreJson(communitySnapshot.data.toMap()),
    );

    // The next lines set text ("statement descriptor") on customer credit card statement line.
    // Max length of full descriptor is 22.
    // The full descriptor is [Shortened descriptor] + '* ' + dynamic suffix.
    // Set [Shortened descriptor] value at https://dashboard.stripe.com/settings/account
    // Length constraint on suffix is therefore (22 - "Community* ".length)
    // There are also character requirements.
    // See https://stripe.com/docs/statement-descriptors

    // Replace non-whitelisted characters with '-'
    final sanitized =
        community.name?.replaceAll(RegExp('[^a-zA-Z0-9 .]'), '-') ?? '';

    // Trim to correct length
    final trimmed = sanitized.substring(0, min(sanitized.length, 15));

    // Replace with generic descriptor if calculated descriptor has no letters
    final descriptor =
        RegExp('^[^a-zA-Z]*\$').hasMatch(trimmed) ? 'Donation' : trimmed;

    PartnerAgreement? agreement;
    final agreementDocs = await firestore
        .collection('partner-agreements')
        .where(
          PartnerAgreement.kFieldCommunityId,
          isEqualTo: communityId,
        )
        .get();
    if (agreementDocs.documents.isNotEmpty) {
      agreement = PartnerAgreement.fromJson(
        firestoreUtils
            .fromFirestoreJson(agreementDocs.documents.first.data.toMap()),
      );
    }

    // use negotiated take rate, if any; otherwise fall back to subscription plan take rate
    final double takeRate = agreement?.takeRate ??
        (await subscriptionPlanUtil.calculateCapabilities(communityId))
            .takeRate!;

    final useConnectedAccount = agreement != null &&
        agreement.allowPayments == true &&
        agreement.stripeConnectedAccountId != null &&
        takeRate >= 0 &&
        takeRate < 1;

    final user =
        (await firestoreUtils.getUsers([context.authUid!].toList())).first;

    final Map<String, String> params = {
      'success_url': 'https://$domain/space/$communityId?donation=success',
      'cancel_url': 'https://$domain/space/$communityId?donation=cancel',
      'payment_method_types[0]': 'card',
      'line_items[0][amount]': '$amountInCents',
      'line_items[0][currency]': 'usd',
      'line_items[0][name]': 'Donation for ${community.name}',
      'line_items[0][quantity]': '1',
      'line_items[0][description]':
          'A portion of this payment may be retained by $orgName per agreement with the recipient organization.',
      'submit_type': 'donate',
      'mode': 'payment',
      'customer': customerId,
      'payment_intent_data[metadata][type]': 'one_time_donation',
      'payment_intent_data[metadata][authUid]': context.authUid!,
      'payment_intent_data[metadata][communityId]': communityId,
      'payment_intent_data[metadata][email]': user.email,
      'payment_intent_data[metadata][name]': user.displayName,
      'payment_intent_data[setup_future_usage]':
          'on_session', // link payment method to customer
      'payment_intent_data[statement_descriptor_suffix]': descriptor,
    };

    if (useConnectedAccount) {
      params['payment_intent_data[transfer_data][destination]'] =
          agreement.stripeConnectedAccountId!;
      params['payment_intent_data[application_fee_amount]'] =
          (amountInCents * takeRate).floor().toString();
    }

    final jsonResponse =
        await stripeUtil.post(path: '/checkout/sessions', params: params);

    final String sessionId = jsonResponse['id'];
    return CreateDonationCheckoutSessionResponse(sessionId: sessionId).toJson();
  }

  // Sets minimum instances to 0
  @override
  void register(FirebaseFunctions functions) {
    functions[functionName] = functions
        .runWith(
          runWithOptions ??
              RuntimeOptions(
                timeoutSeconds: 60,
                memory: '1GB',
                minInstances: 0,
              ),
        )
        .https
        .onCall(callAction);
  }
}

File: firebase/functions/src/admin/payments/create_donation_checkout_session.ts
import * as functions from 'firebase-functions';
import { OnCallMethod } from '../../on_call_function';
import { firestore, firestoreUtils } from '../../utils/infra/firestore_utils';
import { firebaseAuthUtils } from '../../utils/infra/firebase_auth_utils';
import { stripeUtil } from './stripe_util';
import { orElseUnauthorized, orElseNotFound } from '../../utils/utils';

interface CreateDonationCheckoutSessionRequest {
  amountInCents: number;
  communityId: string;
}

export class CreateDonationCheckoutSession extends OnCallMethod<CreateDonationCheckoutSessionRequest> {
  constructor() {
    super(
      'createDonationCheckoutSession',
      (jsonMap) => jsonMap as CreateDonationCheckoutSessionRequest,
      { runWithOptions: { timeoutSeconds: 60, memory: '1GB', minInstances: 0 } }
    );
  }

  async action(
    request: CreateDonationCheckoutSessionRequest,
    context: functions.https.CallableContext
  ): Promise<Record<string, unknown>> {
    const amountInCents = request.amountInCents;
    if (amountInCents <= 0) {
      throw new functions.https.HttpsError('invalid-argument', 'AmountInCents must be greater than zero.');
    }

    const communityId = request.communityId;
    const orgName = functions.config().app?.legal_entity_name as string;

    const customerId = await stripeUtil.getOrCreateCustomerStripeId({ uid: context.auth!.uid });
    const domain = functions.config().app?.domain as string;

    const communitySnap = await firestore.doc(`/community/${communityId}`).get();
    orElseNotFound(communitySnap.exists);
    const community = firestoreUtils.fromFirestoreJson(communitySnap.data() ?? {}) as Record<string, unknown>;

    const sanitized = ((community['name'] as string) ?? '').replace(/[^a-zA-Z0-9 .]/g, '-');
    const trimmed = sanitized.substring(0, Math.min(sanitized.length, 15));
    const descriptor = /^[^a-zA-Z]*$/.test(trimmed) ? 'Donation' : trimmed;

    let agreement: Record<string, unknown> | undefined;
    const agreementDocs = await firestore
      .collection('partner-agreements')
      .where('communityId', '==', communityId)
      .get();
    if (!agreementDocs.empty) {
      agreement = firestoreUtils.fromFirestoreJson(agreementDocs.docs[0].data()) as Record<string, unknown>;
    }

    // Use default take rate = 0 if no agreement
    const takeRate = (agreement?.['takeRate'] as number | undefined) ?? 0;

    const useConnectedAccount =
      agreement != null &&
      agreement['allowPayments'] === true &&
      agreement['stripeConnectedAccountId'] != null &&
      takeRate >= 0 &&
      takeRate < 1;

    const user = (await firebaseAuthUtils.getUsers([context.auth!.uid])).pop();

    const params: Record<string, string> = {
      'success_url': `https://${domain}/space/${communityId}?donation=success`,
      'cancel_url': `https://${domain}/space/${communityId}?donation=cancel`,
      'payment_method_types[0]': 'card',
      'line_items[0][amount]': `${amountInCents}`,
      'line_items[0][currency]': 'usd',
      'line_items[0][name]': `Donation for ${community['name'] ?? ''}`,
      'line_items[0][quantity]': '1',
      'line_items[0][description]': `A portion of this payment may be retained by ${orgName} per agreement with the recipient organization.`,
      'submit_type': 'donate',
      'mode': 'payment',
      'customer': customerId,
      'payment_intent_data[metadata][type]': 'one_time_donation',
      'payment_intent_data[metadata][authUid]': context.auth!.uid,
      'payment_intent_data[metadata][communityId]': communityId,
      'payment_intent_data[metadata][email]': (user as any)?.email ?? '',
      'payment_intent_data[metadata][name]': (user as any)?.displayName ?? '',
      'payment_intent_data[setup_future_usage]': 'on_session',
      'payment_intent_data[statement_descriptor_suffix]': descriptor,
    };

    if (useConnectedAccount) {
      params['payment_intent_data[transfer_data][destination]'] = agreement!['stripeConnectedAccountId'] as string;
      params['payment_intent_data[application_fee_amount]'] = `${Math.floor(amountInCents * takeRate)}`;
    }

    const jsonResponse = await stripeUtil.post({ path: '/checkout/sessions', params });
    return { sessionId: jsonResponse['id'] };
  }
}
